pinot-segment-local/src/main/java/org/apache/pinot/segment/local/upsert/PartitionUpsertMetadataManager.java

```diff
@@ -60,128 +63,170 @@
   private final String _tableNameWithType;
   private final int _partitionId;
   private final ServerMetrics _serverMetrics;
+  private final PartialUpsertHandler _partialUpsertHandler;
 
-  public PartitionUpsertMetadataManager(String tableNameWithType, int partitionId, ServerMetrics serverMetrics) {
+  // TODO(upset): consider an off-heap KV store to persist this index to improve the recovery speed.
```

Comment:
s/upset/upsert/

Reply:
which index? can you be more specific?

Reply:
the index here means PK to recordLocation mapping. cc'ed @Jackie-Jiang i re-phrased this comment a bit.